src/libasr/pass/pass_list_concat.cpp

```diff
@@ -0,0 +1,220 @@
+            do_loop_head.loc = loc;
+            Vec<ASR::stmt_t*> loop_body;
+            loop_body.reserve(al, 1);
+            ASR::stmt_t* list_append = ASRUtils::STMT(ASR::make_ListAppend_t(
+                al, loc, arg_expr, ASRUtils::EXPR(ASR::make_ListItem_t(
+                    al, loc, arg_exprs[0], do_loop_head.m_v, item_type, nullptr))));
+            loop_body.push_back(al, list_append);
+            list_concat_stmt = ASRUtils::STMT(ASR::make_DoLoop_t(
+                al, loc, do_loop_head, loop_body.p, loop_body.n));
+            body.push_back(al, list_concat_stmt);
+
+            // Copy `right_list` contents
+            do_loop_head.m_end = ASRUtils::EXPR(ASR::make_IntegerBinOp_t(
+                al, loc, ASRUtils::EXPR(ASR::make_ListLen_t(
+                    al, loc, arg_exprs[1], int_type, nullptr)
+                ), ASR::binopType::Sub,
+                ASRUtils::EXPR(ASR::make_IntegerConstant_t(
+                    al, loc, 1, int_type)
+                ), int_type, nullptr));
+            loop_body.p = nullptr, loop_body.n = 0;
+            loop_body.reserve(al, 1);
+            list_append = ASRUtils::STMT(ASR::make_ListAppend_t(
+                al, loc, arg_expr, ASRUtils::EXPR(ASR::make_ListItem_t(
+                    al, loc, arg_exprs[1], do_loop_head.m_v, item_type, nullptr))));
+            loop_body.push_back(al, list_append);
+            list_concat_stmt = ASRUtils::STMT(ASR::make_DoLoop_t(
+                al, loc, do_loop_head, loop_body.p, loop_body.n));
+            body.push_back(al, list_concat_stmt);
+
+            ASR::asr_t *fn = ASR::make_Function_t(
+            al, loc,
+            /* a_symtab */ list_concat_symtab,
+            /* a_name */ s2c(al, fn_name),
+            /* a_args */ arg_exprs.p,
+            /* n_args */ arg_exprs.n,
+            /* a_body */ body.p,
+            /* n_body */ body.n,
+            /* a_return_var */ arg_expr,
+            ASR::abiType::Source,
+            ASR::accessType::Public, ASR::deftypeType::Implementation,
+            nullptr,
+            false, false, false, false, false,
+            nullptr, 0,
+            nullptr, 0,
+            false);
+            global_scope->add_symbol(fn_name, down_cast<ASR::symbol_t>(fn));
+            return ASR::down_cast<ASR::symbol_t>(fn);
+        }
+
+        void replace_ListConcat(const ASR::ListConcat_t* x) {
+            Location loc = x->base.base.loc;
+            Vec<ASR::call_arg_t> args;
+            args.reserve(al, 2);
+            ASR::call_arg_t left_list, right_list;
+            left_list.loc = x->m_left->base.loc;
+            left_list.m_value = x->m_left;
+            args.push_back(al, left_list);
+            right_list.loc = x->m_right->base.loc;
+            right_list.m_value = x->m_right;
+            args.push_back(al, right_list);
+            ASR::symbol_t* fn_name = create_concat_function(unit.base.base.loc,
+                unit.m_global_scope, x->m_type);
```

Comment:
Here, I pass `unit.m_global_scope` which I assume is a global_scope.
But in vector copy intrinsic optimization fallback, you pass `current_scope`. So that is the reason for the failure, I think.

Reply:
The `current_scope` is availabe in `PassVisitor` class. I need to assess it somehow, right? Or is there any other way?

Reply:
How do I get the scope in which the `ListConcat` is available?

Reply:
Okay, So I should not worry about the `ListConcat` scope. Why? Because we have to create the function (_lcompilers_list_concat) in the `global_scope`. Then create a new function scope and declare all the variable using that scope. I think this might work

Reply:
I don't know, how to fix this.

Reply:
I got it! So, Our DoLoop is not visited in the BodyVisitor. Instead as @czgdp1807 said it is taken care by the `pass`.